index.js

```diff
@@ -106,54 +121,75 @@ function createFunction(parameters, defaultType) {
     return fun;
 }
 
-function evaluateCategoricalFunction(parameters, input) {
-    for (var i = 0; i < parameters.stops.length; i++) {
-        if (input === parameters.stops[i][0]) {
-            return parameters.stops[i][1];
-        }
+function evaluateCategoricalFunction(parameters, input, hashedStops) {
+    var value = hashedStops[input];
+    if (value === undefined) {
+      // If the input is not found, return the first value from the original array by default
+      return parameters.stops[0][1];
     }
-    return parameters.stops[0][1];
+
+    return value;
 }
 
 function evaluateIntervalFunction(parameters, input) {
-    for (var i = 0; i < parameters.stops.length; i++) {
-        if (input < parameters.stops[i][0]) break;
-    }
-    return parameters.stops[Math.max(i - 1, 0)][1];
+    // Edge cases
+    var n = parameters.stops.length;
+    if (n === 1) return parameters.stops[0][1];
+    if (input <= parameters.stops[0][0]) return parameters.stops[0][1];
+    if (input >= parameters.stops[n - 1][0]) return parameters.stops[n - 1][1];
+
+    var index = binarySearchForIndex(parameters.stops, input);
+
+    return parameters.stops[index][1];
 }
 
 function evaluateExponentialFunction(parameters, input) {
     var base = parameters.base !== undefined ? parameters.base : 1;
 
-    var i = 0;
-    while (true) {
-        if (i >= parameters.stops.length) break;
-        else if (input <= parameters.stops[i][0]) break;
-        else i++;
-    }
-
-    if (i === 0) {
-        return parameters.stops[i][1];
+    // Edge cases
+    var n = parameters.stops.length;
+    if (n === 1) return parameters.stops[0][1];
+    if (input <= parameters.stops[0][0]) return parameters.stops[0][1];
+    if (input >= parameters.stops[n - 1][0]) return parameters.stops[n - 1][1];
```

Comment:
The edge cases are the same code in two places, so we can move it under the `binarySearchForIndex` function to remove repetition.